File: python/benchmarks/xxhash.py
"""The digest boundary against C ``libxxhash``, same protocol, same payload.

Every row is a like-for-like engine comparison: the ``xxhash`` package binds C
``libxxhash``, so a row pairs the same algorithm over the same bytes and the
difference is the engine plus this boundary's conversion cost. The conversion
is deliberately visible rather than averaged away - the ``bytearray`` and
``memoryview`` rows are what a caller pays for a buffer that is not ``bytes``,
and the small-payload rows are where a call's fixed cost still shows.

The final rows measure Arrow holder filling through the Python boundary. They
separate schema insertion, default recomputation, preservation, and ``force``;
the fixture and protocol metadata are built outside every measured call.

The file is named for the topic rather than the module, as ``compression.py``
is: a benchmark named ``xxhash.py`` would shadow the C ``xxhash`` package it
compares against, because a script's own directory comes first on ``sys.path``.

Usage:
    python benchmarks/xxhash.py [--min-time 0.2] [--repeat 5]
"""

from __future__ import annotations

import argparse
import statistics
import sys
import timeit

import pyarrow as pa

from yggdryl import DataType, Field, Scalar, xxhash

try:
    import xxhash as xxhash_c
except ImportError:  # pragma: no cover - the comparison is optional
    xxhash_c = None

PAYLOAD = b'{"id": 1234567, "venue": "XNAS", "price": "150.2500"}\n' * 20_000

#: The sizes where a call's fixed cost, the size branches, and the streaming
#: kernel each dominate in turn.
SIZES = [1, 4, 16, 64, 128, 240, 1024, 64 * 1024, 1024 * 1024]

#: Large enough for per-row work to dominate one Python call.
BATCH_ROWS = 4_096


def _measure(label: str, callable_, size: int, min_time: float, repeat: int) -> None:
    timer = timeit.Timer(callable_)
    number, _ = timer.autorange()
    while timer.timeit(number) < min_time:
        number *= 2
    samples = [timer.timeit(number) / number for _ in range(repeat)]
    median = statistics.median(samples)
    throughput = size / median / (1000 * 1000 * 1000)
    print(f"{label:46s} {median * 1e9:12.1f} ns {throughput:8.2f} GB/s")


def _measure_rows(label: str, callable_, rows: int, min_time: float, repeat: int) -> None:
    timer = timeit.Timer(callable_)
    number, _ = timer.autorange()
    while timer.timeit(number) < min_time:
        number *= 2
    samples = [timer.timeit(number) / number for _ in range(repeat)]
    median = statistics.median(samples)
    throughput = rows / median / 1_000_000
    print(f"{label:46s} {median * 1e6:12.1f} us {throughput:8.2f} M row/s")


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--min-time", type=float, default=0.2)
    parser.add_argument("--repeat", type=int, default=5)
    arguments = parser.parse_args()

    print(f"Python {sys.version.split()[0]}; median of {arguments.repeat}, "
          f"min {arguments.min_time}s per sample")

    cases: list[tuple[str, object, int]] = []

    # One-shot, per size: where the dispatch cost stops mattering.
    for size in SIZES:
        data = PAYLOAD[:size] if size <= len(PAYLOAD) else PAYLOAD * (size // len(PAYLOAD) + 1)
        data = data[:size]
        cases.append((f"xxh3 {size:>9} B", lambda data=data: xxhash.xxh3(data), size))
        if xxhash_c is not None:
            cases.append(
                (
                    f"xxh3 {size:>9} B (C libxxhash)",
                    lambda data=data: xxhash_c.xxh3_64_intdigest(data),
                    size,
                )
            )

    size = len(PAYLOAD)
    # Every algorithm at one size, so the widths compare directly.
    for name, native in (
        ("xxh32", xxhash.xxh32),
        ("xxh64", xxhash.xxh64),
        ("xxh3", xxhash.xxh3),
        ("xxh128", xxhash.xxh128),
    ):
        cases.append((f"{name} payload", lambda native=native: native(PAYLOAD), size))
    if xxhash_c is not None:
        for name, outside in (
            ("xxh32", xxhash_c.xxh32_intdigest),
            ("xxh64", xxhash_c.xxh64_intdigest),
            ("xxh3", xxhash_c.xxh3_64_intdigest),
            ("xxh128", xxhash_c.xxh3_128_intdigest),
        ):
            cases.append(
                (f"{name} payload (C libxxhash)", lambda outside=outside: outside(PAYLOAD), size)
            )

    # The conversion cost, made visible: `bytes` is borrowed, every other
    # buffer is read through one bounded window.
    mutable = bytearray(PAYLOAD)
    view = memoryview(PAYLOAD)
    text = PAYLOAD.decode()
    cases.append(("xxh3 payload (bytearray)", lambda: xxhash.xxh3(mutable), size))
    cases.append(("xxh3 payload (memoryview)", lambda: xxhash.xxh3(view), size))
    cases.append(("xxh3 payload (str)", lambda: xxhash.xxh3(text), size))

    # Streaming against one-shot, and the digest wrapper against the bare int.
    def streamed() -> int:
        state = xxhash.Xxh3()
        for index in range(0, len(PAYLOAD), 64 * 1024):
            state.write_bytes(view[index : index + 64 * 1024])
        return int(state.as_digest())

    cases.append(("xxh3 payload (streamed 64 KiB)", streamed, size))
    cases.append(
        ("digest payload (Digest wrapper)", lambda: xxhash.digest(PAYLOAD, "xxh3-64"), size)
    )

    # The value feed: a leaf, a wide record, and the row a table hashes.
    leaf = Scalar.from_py("AAPL")
    record = Scalar.from_py({f"column_{index:03}": index for index in range(64)})
    cases.append(("scalar leaf digest", lambda: leaf.digest(), 4))
    cases.append(("scalar wide record digest", lambda: record.digest(), 64))
    cases.append(("scalar leaf stable_hash", lambda: leaf.stable_hash(), 4))

    for label, callable_, case_size in cases:
        _measure(label, callable_, case_size, arguments.min_time, arguments.repeat)

    symbols = pa.array(
        ["AAPL" if index % 2 == 0 else "MSFT" for index in range(BATCH_ROWS)]
    )
    holder = Field("row_digest", "uint64", nullable=False)
    holder.digest["role"] = "holder"
    holder.digest["paths"] = '["symbol"]'
    root = Field(
        "row",
        DataType.from_fields([Field("symbol", "utf8", nullable=False), holder]),
        nullable=False,
    )
    missing = pa.record_batch([symbols], names=["symbol"])
    defaults = pa.record_batch(
        [symbols, pa.array([0] * BATCH_ROWS, type=pa.uint64())],
        names=["symbol", "row_digest"],
    )
    populated = pa.record_batch(
        [symbols, pa.array(range(1, BATCH_ROWS + 1), type=pa.uint64())],
        names=["symbol", "row_digest"],
    )
    state = xxhash.Xxh3(seed=7)
    fill_cases = [
        ("fill batch (missing holder)", lambda: state.fill_arrow_batch(root, missing)),
        ("fill batch (default holders)", lambda: state.fill_arrow_batch(root, defaults)),
        ("fill batch (preserve populated)", lambda: state.fill_arrow_batch(root, populated)),
        (
            "fill batch (force populated)",
            lambda: state.fill_arrow_batch(root, populated, force=True),
        ),
    ]
    for label, callable_ in fill_cases:
        _measure_rows(label, callable_, BATCH_ROWS, arguments.min_time, arguments.repeat)

    if xxhash_c is None:
        print("the xxhash package (C libxxhash) is not installed; comparison rows skipped")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: node/benchmarks/xxhash.js
'use strict'

// The digest boundary: every algorithm at the sizes where a call's fixed cost,
// the size branches, and the streaming kernel each dominate in turn, plus the
// conversion cost of each byte shape. The conversion is deliberately visible
// rather than averaged away - a `Buffer` is borrowed, a plain `ArrayBuffer` is
// copied into one before it crosses, and a `string` is encoded. The final rows
// isolate Arrow holder insertion, conditional filling, and force; they include
// the JavaScript boundary's required IPC copies.

const { performance } = require('node:perf_hooks')
const arrow = require('apache-arrow')

const { DataType, Field, Scalar, xxhash } = require('yggdryl')

const PAYLOAD = Buffer.from('{"id": 1234567, "venue": "XNAS", "price": "150.2500"}\n'.repeat(20_000))

/** The sizes a digest's cost changes shape at. */
const SIZES = [1, 4, 16, 64, 128, 240, 1024, 64 * 1024, 1024 * 1024]

/** Large enough for per-row work and the IPC boundary to dominate one call. */
const BATCH_ROWS = 4_096

function measure(name, bytes, iterations, operation) {
  for (let index = 0; index < 5; index += 1) operation()
  const started = performance.now()
  for (let index = 0; index < iterations; index += 1) operation()
  const elapsed = performance.now() - started
  const perOperation = elapsed / iterations
  const throughput = bytes / (perOperation / 1_000) / (1000 * 1000 * 1000)
  console.log(
    `${name.padEnd(44)} ${(perOperation * 1_000_000).toFixed(1).padStart(12)} ns ` +
      `${throughput.toFixed(2).padStart(8)} GB/s`,
  )
}

function measureRows(name, rows, iterations, operation) {
  for (let index = 0; index < 5; index += 1) operation()
  const started = performance.now()
  for (let index = 0; index < iterations; index += 1) operation()
  const elapsed = performance.now() - started
  const perOperation = elapsed / iterations
  const throughput = rows / (perOperation / 1_000) / 1_000_000
  console.log(
    `${name.padEnd(44)} ${(perOperation * 1_000).toFixed(1).padStart(12)} us ` +
      `${throughput.toFixed(2).padStart(8)} M row/s`,
  )
}

function main() {
  console.log(`node ${process.version}`)

  // One-shot, per size: where the dispatch cost stops mattering.
  for (const size of SIZES) {
    const data = Buffer.alloc(size)
    PAYLOAD.copy(data, 0, 0, Math.min(size, PAYLOAD.length))
    const iterations = size > 64 * 1024 ? 2_000 : 200_000
    measure(`xxh3 ${String(size).padStart(9)} B`, size, iterations, () => xxhash.xxh3(data))
  }

  // Every algorithm at one size, so the widths compare directly.
  const size = PAYLOAD.length
  for (const [name, native] of [
    ['xxh32', xxhash.xxh32],
    ['xxh64', xxhash.xxh64],
    ['xxh3', xxhash.xxh3],
    ['xxh128', xxhash.xxh128],
  ]) {
    measure(`${name} payload`, size, 500, () => native(PAYLOAD))
  }

  // The conversion cost, made visible.
  const array = new Uint8Array(PAYLOAD)
  const buffer = array.buffer.slice(0)
  const text = PAYLOAD.toString()
  measure('xxh3 payload (Uint8Array)', size, 500, () => xxhash.xxh3(array))
  measure('xxh3 payload (ArrayBuffer)', size, 500, () => xxhash.xxh3(buffer))
  measure('xxh3 payload (string)', size, 500, () => xxhash.xxh3(text))

  // Streaming against one-shot, and the digest wrapper against the bare bigint.
  measure('xxh3 payload (streamed 64 KiB)', size, 500, () => {
    const state = new xxhash.Xxh3()
    for (let index = 0; index < PAYLOAD.length; index += 64 * 1024) {
      state.writeBytes(PAYLOAD.subarray(index, index + 64 * 1024))
    }
    return state.asDigest().value()
  })
  measure('digest payload (Digest wrapper)', size, 500, () => xxhash.digest(PAYLOAD, 'xxh3-64'))

  // The value feed: a leaf, a wide record, and the hash a table already reads.
  const leaf = Scalar.fromJs('AAPL')
  const record = Scalar.fromJs(
    Object.fromEntries(Array.from({ length: 64 }, (_, index) => [`column_${index}`, BigInt(index)])),
  )
  measure('scalar leaf digest', 4, 200_000, () => leaf.digest())
  measure('scalar wide record digest', 64, 20_000, () => record.digest())
  measure('scalar leaf stableHash', 4, 200_000, () => leaf.stableHash())

  const symbols = arrow.vectorFromArray(
    Array.from({ length: BATCH_ROWS }, (_, index) => (index % 2 === 0 ? 'AAPL' : 'MSFT')),
    new arrow.Utf8(),
  )
  const holder = new Field('row_digest', 'uint64', false, {
    'digest:role': 'holder',
    'digest:paths': '["symbol"]',
  })
  const root = new Field(
    'row',
    DataType.fromFields([new Field('symbol', 'utf8', false), holder]),
    false,
  )
  const missing = new arrow.Table({ symbol: symbols }).batches[0]
  const defaults = new arrow.Table({
    symbol: symbols,
    row_digest: arrow.vectorFromArray(Array(BATCH_ROWS).fill(0n), new arrow.Uint64()),
  }).batches[0]
  const populated = new arrow.Table({
    symbol: symbols,
    row_digest: arrow.vectorFromArray(
      Array.from({ length: BATCH_ROWS }, (_, index) => BigInt(index + 1)),
      new arrow.Uint64(),
    ),
  }).batches[0]
  const signedHolder = new Field('row_digest', 'int64', false, {
    'digest:role': 'holder',
    'digest:paths': '["symbol"]',
  })
  const signedRoot = new Field(
    'row',
    DataType.fromFields([new Field('symbol', 'utf8', false), signedHolder]),
    false,
  )
  const state = new xxhash.Xxh3(7n)
  measureRows('fillArrowBatch (missing holder)', BATCH_ROWS, 100, () => {
    state.fillArrowBatch(root, missing)
  })
  measureRows('fillArrowBatch (missing signed holder)', BATCH_ROWS, 100, () => {
    state.fillArrowBatch(signedRoot, missing)
  })
  measureRows('fillArrowBatch (default holders)', BATCH_ROWS, 100, () => {
    state.fillArrowBatch(root, defaults)
  })
  measureRows('fillArrowBatch (preserve populated)', BATCH_ROWS, 100, () => {
    state.fillArrowBatch(root, populated)
  })
  measureRows('fillArrowBatch (force populated)', BATCH_ROWS, 100, () => {
    state.fillArrowBatch(root, populated, true)
  })
}

main()
